designs/nodepool-requirement-flexibility.md

```diff
@@ -0,0 +1,185 @@
+
+### How does `minValues` interact with consolidation?
+
+The scheduling simulation should return a result from consolidation that respects `minValues` consistent with the launch behavior. Karpenter will then reduce the set of instance types here to be strictly cheaper than the combined price of all instance types that we plan to replace (this is our `filterByPrice` function).
+
+It’s possible that after Karpenter reduces these instance types down to only cheaper ones, it no longer fulfills the `minValues` criteria of the NodePool requirements. As a result, consolidation will need to re-check whether the `minValues` is met after the instance types are filtered down by price to ensure that we can proceed with the launch. If the `minValues` is not satisfied, we should reject the consolidation decision.
+
+### How does `minValues` interact with spot-to-spot consolidation?
+
+[Spot-to-spot consolidation already requires a spot instance flexibility of 15](https://github.com/kubernetes-sigs/karpenter/blob/main/designs/spot-consolidation.md). If you do not specify any `minValues` across any of your requirements in your NodePool, that 15 instance type minimum will still be required and fulfilled.
+
+If a user has `minValues` set in their NodePool that extend the number of instance types that would be needed for a spot-to-spot consolidation beyond 15, that number of instance types would be maintained. For instance, if a user had the following requirements:
+
+```
+- key: node.kubernetes.io/instance-type
+  operator: Exists
+  minValues: 30
+```
+
+In this case, we would maintain 30 instance type values in the launch (rather than truncating down to 15), assuming that the scheduling simulation succeeds. Since 30 > 15 instance types that we require to do spot-to-spot consolidation.
+
+If a user has `minValues` set in their NodePool that require less than 15 instance types, then the 15 instance type minimum will continue to be enforced to enable a spot-to-spot consolidation to occur. For instance, if a user had the following requirements:
+
+```
+- key: node.kubernetes.io/instance-type
+  operator: Exists
+  minValues: 3
+```
+
+In this case, we would schedule with at least 3 instance types and then validate that the result allowed for 15 cheaper instance types that we could consolidate to. If there were not 15 cheaper instance types, then we would not perform spot-to-spot consolidation.
```

Comment:
Just sanity check here: Are you saying that we'd check if we can consolidate first and see if there are 15 instance types, and then after the fact, check to make sure that at least 3 of those instance types are within the requirements of the nodePool? Sounds like we'd have to add a flexibility check during the scheduling simulation and a flexibility check after the results parsing

Reply:
That's what I'm saying, yes. For consolidation, because we filter out cheaper instance types, we have to perform the scheduling simulation, then filter out the cheaper instance types from the result, then check that we still have 15 instance types and that we are maintaining the instance type flexibility here.